Shutdown race condition in PersistentExecutorImpl causing IllegalStateException

**Describe the bug**  
A race condition exists in PersistentExecutorImpl where background polling tasks (FinderTask) continue running during server shutdown after database resources are closed, resulting in IllegalStateException: Attempting to execute an operation on a closed EntityManager.
```
[2/27/26, 13:08:16:825 UTC] 000000ee com.ibm.ws.concurrent.internal.ScheduledTask                 E CWWKC1101E: The task test.concurrent.persistent.fat.simctrl.SimulatedController$FinderTask@d77d915b, which was submitted to executor service managedScheduledExecutorService[DefaultManagedScheduledExecutorService], failed with the following error: javax.management.RuntimeMBeanException: java.lang.IllegalStateException
	at java.management/com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.rethrow(DefaultMBeanServerInterceptor.java:819)
	at java.management/com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.rethrowMaybeMBeanException(DefaultMBeanServerInterceptor.java:832)
	at java.management/com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.invoke(DefaultMBeanServerInterceptor.java:805)
	at java.management/com.sun.jmx.mbeanserver.JmxMBeanServer.invoke(JmxMBeanServer.java:802)
	at com.ibm.ws.kernel.boot.jmx.service.MBeanServerForwarderDelegate.invoke(MBeanServerForwarderDelegate.java:198)
	at com.ibm.ws.jmx.internal.DelayedMBeanActivator.invoke(DelayedMBeanActivator.java:516)
	at com.ibm.ws.kernel.boot.jmx.internal.PlatformMBeanServer.invoke(PlatformMBeanServer.java:374)
	at test.concurrent.persistent.fat.simctrl.SimulatedController$FinderTask.call(SimulatedController.java:213)
	at test.concurrent.persistent.fat.simctrl.SimulatedController$FinderTask.call(SimulatedController.java:186)
	at com.ibm.ws.concurrent.internal.ScheduledTask.call(ScheduledTask.java:450)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1595)
Caused by: java.lang.IllegalStateException
	at com.ibm.ws.concurrent.persistent.db.DatabaseTaskStore.getPersistenceServiceUnit(DatabaseTaskStore.java:921)
	at com.ibm.ws.concurrent.persistent.db.DatabaseTaskStore.find(DatabaseTaskStore.java:372)
	at com.ibm.ws.concurrent.persistent.internal.PersistentExecutorMBeanImpl.buildAndLogException(PersistentExecutorMBeanImpl.java:190)
	at com.ibm.ws.concurrent.persistent.internal.PersistentExecutorMBeanImpl.findTaskIds(PersistentExecutorMBeanImpl.java:120)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:586)
	at sun.reflect.misc.Trampoline.invoke(MethodUtil.java:64)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:586)
	at java.base/sun.reflect.misc.MethodUtil.invoke(MethodUtil.java:97)
	at java.management/com.sun.jmx.mbeanserver.StandardMBeanIntrospector.invokeM2(StandardMBeanIntrospector.java:112)
	at java.management/com.sun.jmx.mbeanserver.StandardMBeanIntrospector.invokeM2(StandardMBeanIntrospector.java:46)
	at java.management/com.sun.jmx.mbeanserver.MBeanIntrospector.invokeM(MBeanIntrospector.java:236)
	at java.management/com.sun.jmx.mbeanserver.PerInterface.invoke(PerInterface.java:138)
	at java.management/com.sun.jmx.mbeanserver.MBeanSupport.invoke(MBeanSupport.java:252)
	at java.management/javax.management.StandardMBean.invoke(StandardMBean.java:405)
	at java.management/com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.invoke(DefaultMBeanServerInterceptor.java:803)

```
**Steps to Reproduce**  
This issue reported in a build break, it's an intermittent issue
**Expected behavior**  
The error should not be thrown


**Diagnostic information:**  
OpenLiberty Version
26.0.0.3 (wlp-1.0.111.cl260320260226-0302)

Build Date: February 26, 2026
Full version string: WebSphere Application Server 26.0.0.3 (wlp-1.0.111.cl260320260226-0302)
Affected Feature(s)
persistentExecutor-1.0 - Primary affected feature
concurrent-1.0 - Related feature
jdbc-4.1 - Database connectivity
jndi-1.0 - JNDI lookup support
servlet-3.1 - Web application support
usr:simulatedController-1.0 - Custom test feature
controllerTestFeature-1.0 - Test helper feature

Java Version:
java version "21.0.9"
IBM Semeru Runtime Open Edition (21.0.9+10-LTS)
Eclipse OpenJ9 VM

Full details:
Version: 21.0.9+10-LTS
Vendor: IBM Semeru Runtime Open Edition
VM: Eclipse OpenJ9 VM
Build Date: 2025-10-21

Database Configuration:
Type: Apache Derby 10.11.1.1 (Embedded)
Driver: Apache Derby Embedded JDBC Driver 10.11.1.1
Database: In-memory (memory:persistctrldb)
Create on startup: Yes

Operating System
OS: Windows 11 (10.0; amd64)

Error Details
Primary Error:

Exception: java.lang.IllegalStateException
Message: "Attempting to execute an operation on a closed EntityManager"
Location: PersistentExecutorImpl.java:671
Timestamp: 2/27/26, 13:08:15:659 UTC
Secondary Error (during rollback):

Exception: java.sql.SQLNonTransientConnectionException
Message: "No current connection"
Derby Error Code: ERROR 08003
Location: WSRdbSpiLocalTransactionImpl.java:192
Timestamp: 2/27/26, 13:08:15:994 UTC

--- a/dev/com.ibm.ws.concurrent.persistent/src/com/ibm/ws/concurrent/persistent/internal/PersistentExecutorImpl.java
+++ b/dev/com.ibm.ws.concurrent.persistent/src/com/ibm/ws/concurrent/persistent/internal/PersistentExecutorImpl.java
@@ -309,6 +309,12 @@ public class PersistentExecutorImpl implements ApplicationRecycleComponent, DDLG
     private final AtomicBoolean pollingStartSignalReceived = new AtomicBoolean();
 
     /**
+     * Indicates if the polling task is currently executing.
+     * Used to prevent race condition during deactivation.
+     */
+    private final AtomicBoolean pollingTaskRunning = new AtomicBoolean(false);
+
+    /**
      * Liberty scheduled executor.
      */
     @Reference(target = "(deferrable=false)")
@@ -507,6 +513,25 @@ public class PersistentExecutorImpl implements ApplicationRecycleComponent, DDLG
      */
     protected void deactivate(ComponentContext context) throws Exception {
         deactivated = true;
+        
+        // Cancel the polling future to stop background tasks before closing resources 
+        // product bug fix 308877
+        ScheduledFuture<?> pollingFuture = pollingFutureRef.get();
+        if (pollingFuture != null) {
+            pollingFuture.cancel(false);
+            // Wait for the polling task to actually finish to avoid race condition with resource cleanup
+            // Cannot use get() or isDone() on cancelled future, so check our own flag
+            long deadline = System.nanoTime() + TimeUnit.SECONDS.toNanos(5);
+            while (pollingTaskRunning.get() && System.nanoTime() < deadline) {
+                try {
+                    Thread.sleep(50);
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                    break;
+                }
+            }
+        }
+        
         if (mbean != null) {
             mbean.unregister();
             mbean = null;
@@ -2593,7 +2618,10 @@ public class PersistentExecutorImpl implements ApplicationRecycleComponent, DDLG
             if (trace && tc.isEntryEnabled())
                 Tr.entry(PersistentExecutorImpl.this, tc, "run[poll]");
 
-            Config config = configRef.get();
+            // Claim execution state BEFORE checking deactivated to prevent race condition
+            pollingTaskRunning.set(true);
+            try {
+                Config config = configRef.get();
 
             if (deactivated || !config.enableTaskExecution || config != initialConfig) {
                 if (trace && tc.isEntryEnabled())
@@ -2638,6 +2666,9 @@ public class PersistentExecutorImpl implements ApplicationRecycleComponent, DDLG
 
             if (trace && tc.isEntryEnabled())
                 Tr.exit(PersistentExecutorImpl.this, tc, "run[poll]", failure);
+            } finally {
+                pollingTaskRunning.set(false);
+            }
         }
     }
 

